### PR TITLE
[move-compiler-v2] fix issue 6922 by allowing type annotations for lambda parameters

### LINKED ISSUE #6922: [Feature Request] Support type annotations for lambda parameters
Currently it is not possible to provide type annotations for lambda parameters. Instead of `|x: T| e` one has to write `|x| { let x:T = x; e }`. Type annotations are required to help the inference algorithm in some occasions.

### PR DESCRIPTION
**Pull request by brmataptos**
## Description

Allow type annotations on lambda parameters in Move 2.

Fixes #6922.

## How Has This Been Tested?

Added type annotations to lambda params on a lot of existing tests, renamed as `*_typed.move`.
Did a diff of outputs to make sure outputs are similar to the previous, without annotations, where appropriate.

Later tests of more interesting Lambda uses will test more subtyping cases before release.

## Key Areas to Review

This includes some cases that should involve integer/u64 subtyping,
but maybe there could be more interesting subtyping cases checked.
Feel free to suggest a feasible test case.

Someone more familiar with the type widening directions than I should take a look at the spot where
I have a comment "is this arg ordering right?" in exp_builder.rs.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-compiler-v2/transactional-tests/tests/inlining/bug_10991_typed.exp [hunk 1]
```diff
@@ -0,0 +1,41 @@
+comparison between v1 and v2 failed:
+- processed 2 tasks
+- 
+- task 0 'publish'. lines 1-10:
+- Error: error[E01013]: unsupported language construct
+-   ┌─ TEMPFILE:7:22
+-   │
+- 7 │         assert!(foo(|x: u64, _: u64| x, |_: u64, y: u64| y, 10, 100) == 110, 0);
+-   │                      ^^^^^^ Typed parameter to lambda is only allowed in Move 2 and beyond
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/transactional-tests/tests/inlining/bug_10991_typed.exp:9]: nit: could we reword this to - "Explicit type annotations for lambda parameters are only allowed ..."

Because the lambda parameters are indeed typed even in Move 1, just that explicit type annotations are not allowed.

File: third_party/move/move-compiler/src/expansion/ast.rs [hunk 2]
```diff
@@ -421,6 +421,12 @@ pub type LValue = Spanned<LValue_>;
 pub type LValueList_ = Vec<LValue>;
 pub type LValueList = Spanned<LValueList_>;
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct TypedLValue_(pub LValue, pub Option<Type>);
```
Comment (vineethk) [inline third_party/move/move-compiler/src/expansion/ast.rs:425]: Perhaps add a brief comment here saying these are used to represent explicitly typed (by the developer) lvalues.

File: third_party/move/move-compiler/src/expansion/translate.rs [hunk 3]
```diff
@@ -2883,6 +2883,38 @@ fn fields<T>(
...
+}
+
+fn typed_bind(context: &mut Context, sp!(loc, tpb_): P::TypedBind) -> Option<E::TypedLValue> {
+    let P::TypedBind_(pb, opt_type) = tpb_;
+    let b = bind(context, pb)?;
+    let ot = opt_type.map(|ty| type_(context, ty));
+    Some(sp(loc, E::TypedLValue_(b, ot)))
+}
+
+// Some(E::TypedLValue_(sp(loc, b_));
```
Comment (vineethk) [inline third_party/move/move-compiler/src/expansion/translate.rs:2904]: Looks like some leftover code that should be removed.

File: third_party/move/move-compiler/src/parser/ast.rs [hunk 4]
```diff
@@ -518,6 +518,12 @@ pub type Bind = Spanned<Bind_>;
 // b1, ..., bn
 pub type BindList = Spanned<Vec<Bind>>;
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct TypedBind_(pub Bind, pub Option<Type>);
```
Comment (vineethk) [inline third_party/move/move-compiler/src/parser/ast.rs:522]: We should have a test case where some parameters of a lambda are explicitly typed, but some are not (e.g., `|x: u64, y| { x + y }`). Do we want this to be legal?

File: third_party/move/move-compiler/src/parser/ast.rs [hunk 5]
```diff
@@ -658,7 +664,7 @@ pub enum Exp_ {
     // { seq }
     Block(Sequence),
     // |x1, ..., xn| e
```
Comment (vineethk) [inline third_party/move/move-compiler/src/parser/ast.rs:666]: nit: fix this comment to include explicit typing of parameters?

File: third_party/move/move-compiler-v2/tests/checking/typing/eq_inline_typed.move [hunk 6]
```diff
@@ -0,0 +1,13 @@
+module 0x42::m {
+
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/checking/typing/eq_inline_typed.move:2]: The referenced issue #6922 says "type annotations are required to help the inference algorithm in some occasions". Do we have an example/test case where the type inference was not able to infer the types when lambda params were not explicitly typed, but the addition of the feature in this PR allows the user to progress in the natural way by adding explicit types (instead of the hack mentioned in the referenced issue for explicit typing)?

File: third_party/move/move-model/src/builder/exp_builder.rs [hunk 7]
```diff
@@ -2328,6 +2383,33 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
         }
     }
```
Comment (vineethk) [inline third_party/move/move-model/src/builder/exp_builder.rs:2385]: It would be good to add some test cases where lambda params and return types involve some automatic type coercion (between `&mut` and `&`) due to subtyping:
- acceptable "mismatch" between expected and provided types (automatic type coercion makes it okay to have these mismatches), on the return type and param type of a lambda (the directionality of the acceptable mismatch is different for these)
- erroneous mismatch between expected and provided types on the return type and param types of a lambda.

**Review by vineethk — COMMENTED** [review]
Looking pretty good so far. Have a few comments.

File: third_party/move/move-compiler-v2/tests/checking-lang-v1/eq_inline_typed.move [hunk 8]
```diff
@@ -0,0 +1,13 @@
+module 0x42::m {
+
+    inline fun foo(f: |&u64|) {
+    }
+
+    fun g() {
+        foo(|v: &u64| {
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/checking-lang-v1/eq_inline_typed.move:7]: Once this PR is in, we should fix the text in the move book: https://aptos.dev/en/build/smart-contracts/book/functions#current-restrictions, where we currently say this is not allowed.

File: third_party/move/move-model/src/builder/exp_builder.rs [hunk 9]
```diff
@@ -2328,6 +2383,33 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
...
+                let loc = self.to_loc(loc);
+                let bound_type = self.translate_type(ty);
+                self.check_type_with_order(
+                    expected_order,
+                    &loc,
+                    &bound_type, // is this arg ordering right?
+                    expected_type,
+                    context,
+                );
+                self.translate_lvalue(lv, &bound_type, expected_order, match_locals, context)
```
Comment (fEst1ck) [inline third_party/move/move-model/src/builder/exp_builder.rs:2407]: Why are we using the `expected_order` here? Shouldn't we check tha lvalues have the `bound_type` exactly, similar to how wo check annotated expressions?
```
EA::Exp_::Annotate(exp, typ) => {
	let ty = self.translate_type(typ);
	let exp =
		self.translate_exp_in_context(exp, &ty, &ErrorMessageContext::TypeAnnotation);
	self.check_type(&loc, &ty, expected_type, context);
	exp
},
```

File: third_party/move/move-compiler/src/parser/ast.rs (hunk 4 above)
Comment (brmataptos) [reply third_party/move/move-compiler/src/parser/ast.rs:522]: Added a test case at the end of `third_party/move/move-compiler-v2/tests/checking/typing/lambda_typed.move` and new test `third_party/move/move-compiler-v2/transactional-tests/tests/inlining/eval_ignored_param_some_typed.move`.

File: third_party/move/move-compiler-v2/tests/checking/typing/eq_inline_typed.move (hunk 6 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tests/checking/typing/eq_inline_typed.move:2]: I'm still trying to find such a case.

File: third_party/move/move-compiler/src/expansion/translate.rs (hunk 3 above)
Comment (brmataptos) [reply third_party/move/move-compiler/src/expansion/translate.rs:2904]: Done

File: third_party/move/move-compiler/src/expansion/ast.rs (hunk 2 above)
Comment (brmataptos) [reply third_party/move/move-compiler/src/expansion/ast.rs:425]: Done

File: third_party/move/move-compiler-v2/transactional-tests/tests/inlining/bug_10991_typed.exp (hunk 1 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/transactional-tests/tests/inlining/bug_10991_typed.exp:9]: Done.

File: third_party/move/move-model/src/builder/exp_builder.rs (hunk 9 above)
Comment (brmataptos) [reply third_party/move/move-model/src/builder/exp_builder.rs:2407]: I don't think so.  The type of the `LValue` is the type of a container, not a value.  It may be wider than the value put in it.  I've added a test `lambda_typed_widen.move` which tries to use the "subtyping" relationship between `&mut u64` and `& u64` to test what happens to lambda params with declared types versus those with undeclared types.  Not sure it's a strong enough test, though.

File: third_party/move/move-model/src/builder/exp_builder.rs (hunk 7 above)
Comment (brmataptos) [reply third_party/move/move-model/src/builder/exp_builder.rs:2385]: I've added `lambda_typed_widen.move` that tries to do something with that, but I can't think of any subtyping relationship we have other than `&` and `&mut`.  Any better ideas?

File: third_party/move/move-compiler-v2/tests/checking-lang-v1/eq_inline_typed.move (hunk 8 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/tests/checking-lang-v1/eq_inline_typed.move:7]: @brmataptos Just double checking this comment was seen.

File: third_party/move/move-model/src/builder/exp_builder.rs (hunk 7 above)
Comment (vineethk) [reply third_party/move/move-model/src/builder/exp_builder.rs:2385]: I believe that is the only subtyping relationship we currently have.

**Review by vineethk — APPROVED** [review]

**Review by brmataptos — COMMENTED** [review]
I think I found all of your comments, Vineeth.  Please let me know if not.

File: third_party/move/move-compiler-v2/tests/checking/typing/lambda_typed_widen.move [hunk 10]
```diff
@@ -0,0 +1,51 @@
+module 0x8675309::M {
+    use std::vector;
+
+    public inline fun use_mut_ref<T>(v: &mut vector<T>, action: |&mut T|T): T {
```
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/tests/checking/typing/lambda_typed_widen.move:4]: Could you write a test case where the return value of the lambda function is immutable/mutable reference?

Comment (brmataptos) [reply third_party/move/move-compiler-v2/tests/checking/typing/lambda_typed_widen.move:4]: The new tests do reveal some awkward error messages, which can be elicited without this PR, please see #14681.

**Review by rahxephon89 — APPROVED** [review]
